keps/sig-cluster-lifecycle/draft-20180130-kubeadm-join-master.md

```diff
@@ -0,0 +1,429 @@
+  - "@luxas"
+  - "@timothysc"
+  - "@roberthbailey"
+editor:
+  - "@fabriziopandini"
+creation-date: 2018-01-28
+last-updated: 2018-01-28
+see-also:
+  - KEP 0004
+
+```
+
+## Table of Contents
+
+  * [kubeadm join --master workflow](#kubeadm-join---master-workflow)
+    * [Metadata](#metadata)
+    * [Table of Contents](#table-of-contents)
+    * [Summary](#summary)
+    * [Motivation](#motivation)
+      * [Goals](#goals)
+      * [Non-goals](#non-goals)
+      * [Challenges and Open Questions](#challenges-and-open-questions)
+    * [Proposal](#proposal)
+      * [User Stories](#user-stories)
+        * [Add a new master node](#add-a-new-master-node)
+      * [Implementation Details](#implementation-details)
+        * [advertise-address = IP/DNS of the external load balancer](#advertise-address--ipdns-of-the-external-load-balancer)
+        * [kubeadm init --feature-gates=HighAvailability=true](#kubeadm-init---feature-gateshighavailabilitytrue)
+        * [kubeadm join --master workflow](#kubeadm-join---master-workflow-1)
+        * [Strategies for deploying control plane components](#strategies-for-deploying-control-plane-components)
+        * [Strategies for distributing cluster certificates](#strategies-for-distributing-cluster-certificates)
+    * [Graduation Criteria](#graduation-criteria)
+    * [Implementation History](#implementation-history)
+    * [Drawbacks](#drawbacks)
+    * [Alternatives](#alternatives)
+
+## Summary
+
+We are extending the kubeadm distinctive `init` and `join` workflow, introducing the
+capability to add more than one master node to an existing cluster by means of the
+new `kubeadm join --master` option.
+
+As a consequence, kubeadm will provide a best-practice, “fast path” for creating a
+minimum viable, conformant Kubernetes cluster with one or more master nodes and
+zero or more worker nodes.
+
+## Motivation
+
+Support for high availability is one of the most requested features for kubeadm.
+
+Even if, as of today, there is already the possibility to create an HA cluster
+using kubeadm in combination with some scripts and/or automation tools (e.g.
+[this](https://goo.gl/yStTLs)), this KEP was designed with the objective to introduce
+an "up-stream" simple and reliable solution for achieving the same goal.
+
+### Goals
+
+* "Divide and conquer”
+
+  This proposal - at least in its initial release - does not address all the possible
+  user stories for creating an highly available Kubernetes cluster, but instead
+  focuses on:
+
+  * Defining a generic and extensible flow for bootstrapping an HA cluster, the
+    `kubeadm join --master` workflow.
+  * Providing a solution *only* for one, well defined user story. see
+    [User Stories](#user-stories) and [Non-goals](#non-goals).
+
+* Provide support for a dynamic bootstrap flow
+
+  At the time a user is running `kubeadm init`, s/he/an operator might not know what
+  the cluster setup will look like eventually. For instance, the user may start with
```

Comment:
I am not sure why this has to be the case, it's seems like more of nice-to-have, in contrast to providing simple path of getting an HA cluster with reliable and deterministic bootstrap process.

Reply:
IMO if we focus too much on a deterministic bootstrap process we are implicitly failing in addressing many other use cases and most importantly we are failing to address what can happen during the cluster lifecycle e.g. a master node should be replaced 
As a consequence, I consider the dynamic workflow bootstrap flow an important part of the value that kubeadm can bring to HA in Kubernestes; I propose to discuss this as well in the next kubeadm office hours meeting